store/memdb/memdb.go

```diff
@@ -333,3 +343,63 @@ RunLoop:
 
 	return nil, nil
 }
+
+// UpsertSHAPair creates a sha pair record for a pipeline id.
+func (m *MemDB) UpsertSHAPair(pair gaia.SHAPair) error {
+	// Create a write transaction
+	txn := m.db.Txn(true)
+
+	// Find existing entry
+	raw, err := txn.First(shaPairTable, "id", pair.UniqueID)
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to lookup sha pair record via upsert", "error", err.Error())
+		return err
+	}
+
+	// Delete if it exists
+	if raw != nil {
+		err = txn.Delete(shaPairTable, raw)
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed to delete sha pair via upsert", "error", err.Error())
+			return err
+		}
+	}
+
+	// Insert it
+	if err := txn.Insert(shaPairTable, pair); err != nil {
+		gaia.Cfg.Logger.Error("failed to insert sha pair via upsert", "error", err.Error())
+		return err
+	}
+
+	// Commit transaction
+	txn.Commit()
+	return nil
+}
+
+// GetSHAPair returns a sha pair for a pipeline id.
+func (m *MemDB) GetSHAPair(pipelineID string) (found bool, pair gaia.SHAPair, err error) {
```

Comment:
Wouldn't it be easier to return a pointer and return a nil-pointer in case we couldn't find it?

Reply:
There is no point in sharing this struct, so, no. :) I don't see a "point" :D in returning a pointer. I would rather use the `bool, struct, error` pattern in this case. The less pointers the application is using the better the GC will be.

Reply:
When we load the data from the memdb we create a new struct anyway. The source routine which is calling this function is the only function which will own this struct. So this would actually increase the performance since you don't have to copy again the struct, right?

Reply:
As long as I don't see benchmarks for that, that's an assumption. :)

If there is no point in sharing a data variable I won't share it. If I don't know what a copy does to a struct I will use a pointer, otherwise, value semantics should also be a first choice. Especially not up the call stack.

Reply:
To be honest, I'm still confused about best practices around value/pointer usage. Some people say you should usually use values others say use whatever you want and then others say just be strict within your once code-base decision. 😞
If you look at the go standard libraries it is also mixed up quite heavy.

Reply:
I think the standards are: 
- use value by default because it's safe and helps out the GC
- use pointers for sharing ( but only downwards avoid upwards sharing )
- use pointers if you definitely have a case where it provides a performance boost
- use pointers if you aren't sure if a value can be copied ( for example an unknown struct type or a reference / interface type ).